File: import_admins_to_firestore.py
#!/usr/bin/env python3
# import_admins_to_firestore.py
import argparse, json, sys
from typing import Dict, Any, Optional, Iterable

import firebase_admin
from firebase_admin import credentials, firestore

BATCH_LIMIT = 450  # jätetään marginaali Firestoren 500/doc batch-rajaan

def init_firestore(project_id: Optional[str], creds_path: Optional[str]):
    if firebase_admin._apps:
        return firestore.client()
    if creds_path:
        cred = credentials.Certificate(creds_path)
        firebase_admin.initialize_app(cred, {"projectId": project_id} if project_id else None)
    else:
        firebase_admin.initialize_app(options={"projectId": project_id} if project_id else None)
    return firestore.client()

def pick_id(props: Dict[str, Any]) -> int:
    for k in ("@id", "osm_id", "id"):
        if k in props and props[k] is not None:
            try:
                return int(props[k])
            except Exception:
                return int(str(props[k]).strip())
    raise ValueError("OSM id puuttuu (@id/osm_id/id). Aja osmium export -a id,type.")

def pick_level(props: Dict[str, Any]) -> Optional[int]:
    al = str(props.get("admin_level", "")).strip()
    return int(al) if al in {"2", "4", "8"} else None

def extract_names(props: Dict[str, Any], keep_langs: Optional[set]) -> Dict[str, str]:
    names = {}
    base = props.get("name")
    if isinstance(base, str) and base.strip():
        names["name"] = base.strip()
    for k, v in props.items():
        if isinstance(k, str) and isinstance(v, str) and k.startswith("name:") and v.strip():
            lang = k.split(":", 1)[1]
            if keep_langs is None or lang in keep_langs:
                names[lang] = v.strip()
    return names

def choose_display_name(ns: Dict[str, str]) -> Optional[str]:
    return ns.get("en") or ns.get("fi") or ns.get("name") or (next(iter(ns.values())) if ns else None)

def parse_geojson(path: str, keep_langs: Optional[Iterable[str]]) -> Dict[str, Any]:
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
    feats = data.get("features", [])
    keep_langs = set(keep_langs) if keep_langs else None

    docs: Dict[str, Any] = {}
    for feat in feats:
        props = feat.get("properties") or {}
        geom  = feat.get("geometry") or {}
        if props.get("@type") != "relation":
            continue
        if geom.get("type") not in ("Polygon", "MultiPolygon"):
            continue
        level = pick_level(props)
        if level is None:
            continue
        oid = pick_id(props)
        names = extract_names(props, keep_langs)
        disp  = choose_display_name(names)

        key = f"l{level}_{oid}"
        if key in docs:
            # yhdistä nimiä jos sama id esiintyy useasti
            docs[key]["names"].update(names)
            if disp and not docs[key].get("displayName"):
                docs[key]["displayName"] = disp
        else:
            payload = {"level": level, "id": oid, "names": names}
            if disp:
                payload["displayName"] = disp
            docs[key] = payload
    return docs

def parse_admins_map(path: str) -> Dict[str, Any]:
    # odottaa muotoa { "l8_123": {level, id, names, ...}, ... }
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
    if not isinstance(data, dict):
        raise ValueError("admins-json ei ole dict.")
    # varmista että jokaisella on level/id
    out = {}
    for k, v in data.items():
        if not isinstance(v, dict):
            continue
        if "level" not in v or "id" not in v:
            # yritä päätellä avaimesta
            try:
                lvl_str, oid_str = k.split("_", 1)
                lvl = int(lvl_str[1:])
                oid = int(oid_str)
                v = {"level": lvl, "id": oid, **v}
            except Exception:
                continue
        out[k] = v
    return out

def load_docs(path: str, keep_langs: Optional[Iterable[str]]) -> Dict[str, Any]:
    # Päätä formaatti: GeoJSON jos top-levelissa on "features", muuten oletetaan admins-map
    with open(path, "r", encoding="utf-8") as f:
        head = json.load(f)
    # lue uudelleen oikealla parserilla (palautetaan kursorin alkuun ei kannata kikkailla)
    if isinstance(head, dict) and "features" in head:
        # kirjoita tilapäisesti takaisin ja parse_geojson avaa itse uudelleen
        with open(path, "w", encoding="utf-8") as f2:
            json.dump(head, f2)
        return parse_geojson(path, keep_langs)
    else:
        # kirjoita takaisin ja käsittele map-muotona
        with open(path, "w", encoding="utf-8") as f2:
            json.dump(head, f2)
        return parse_admins_map(path)

def write_in_batches(db, collection: str, items: Dict[str, Any], dry_run: bool=False) -> int:
    col = db.collection(collection)
    count = 0
    in_batch = 0
    batch = db.batch()
    for doc_id, payload in items.items():
        if dry_run:
            count += 1
            continue
        ref = col.document(doc_id)
        batch.set(ref, payload, merge=True)
        in_batch += 1
        count += 1
        if in_batch >= BATCH_LIMIT:
            batch.commit()
            print(f"Committed {in_batch} docs (total {count})")
            batch = db.batch()
            in_batch = 0
    if not dry_run and in_batch > 0:
        batch.commit()
        print(f"Committed {in_batch} docs (total {count})")
    return count

def main():
    ap = argparse.ArgumentParser(description="Import admins (GeoJSON or admins.json) -> Firestore admin_areas/")
    ap.add_argument("--admins-json", required=True, help="Path to admin-248.geojson OR admins.json")
    ap.add_argument("--project", help="GCP project id")
    ap.add_argument("--creds", help="Service account JSON")
    ap.add_argument("--keep-langs", help="Comma-separated lang codes to keep (e.g. fi,sv,en)")
    ap.add_argument("--dry-run", action="store_true", help="Parse only, don't write")
    args = ap.parse_args()

    keep_langs = [x.strip() for x in args.keep_langs.split(",")] if args.keep_langs else None

    try:
        docs = load_docs(args.admins_json, keep_langs)
    except Exception as e:
        print(f"ERROR parsing {args.admins_json}: {e}", file=sys.stderr)
        sys.exit(2)

    print(f"Docs ready to import: {len(docs)}")

    db = init_firestore(args.project, args.creds)
    total = write_in_batches(db, "admin_areas", docs, dry_run=args.dry_run)
    print(("DRY RUN - " if args.dry_run else "") + f"Imported {total} docs into 'admin_areas/'")

if __name__ == "__main__":
    main()

File: import_admins_to_firestore.js
#!/usr/bin/env node
/* eslint-disable no-console */

//
// import_admins_to_firestore.js
//
// Käyttö:
//   node import_admins_to_firestore.js \
//     --admins-json admin-248.geojson \
//     --project your-project-id \
//     --creds serviceAccountKey.json [--dry-run]
//
// Vaatii:
//   npm install firebase-admin stream-json
//
// Huom:
//  - GeoJSON luetaan streamaten (ei JSON.parse koko tiedostoon)
//  - Firestoreen kirjoitetaan batch kerrallaan (BATCH_LIMIT < 500)
//  - names.* kirjoitetaan pistemuotoisina kenttinä (merge täydentää)
//

const fs = require("fs");
const path = require("path");
const admin = require("firebase-admin");

// Stream-json
const { chain } = require("stream-chain");
const { parser } = require("stream-json");
const { pick } = require("stream-json/filters/Pick");
const { streamArray } = require("stream-json/streamers/StreamArray");

const BATCH_LIMIT = 450; // jätetään selvä marginaali Firestoren 500 rajasta

// ------------------ CLI ------------------
function parseArgs(argv) {
    const out = {};
    for (let i = 2; i < argv.length; i++) {
        const k = argv[i];
        const v = argv[i + 1];
        if (k === "--admins-json") { out.input = v; i++; }
        else if (k === "--project") { out.project = v; i++; }
        else if (k === "--creds") { out.creds = v; i++; }
        else if (k === "--dry-run") { out.dryRun = true; }
        else {
            console.error(`Unknown arg: ${k}`);
            process.exit(2);
        }
    }
    if (!out.input) {
        console.error("Usage: node import_admins_to_firestore.js --admins-json <file> [--project <id>] [--creds <sa.json>] [--dry-run]");
        process.exit(1);
    }
    return out;
}
const args = parseArgs(process.argv);

// ---------------- Firestore init ----------------
function initFirestore(projectId, credsPath) {
    if (!admin.apps.length) {
        let cred;
        if (credsPath) {
            const abs = path.resolve(credsPath);
            cred = admin.credential.cert(JSON.parse(fs.readFileSync(abs, "utf8")));
        } else {
            cred = admin.credential.applicationDefault();
        }
        const opt = {};
        if (projectId) opt.projectId = projectId;
        admin.initializeApp({ credential: cred, ...opt });
    }
    return admin.firestore();
}

// ---------------- Helpers ----------------
function pickId(props) {
    for (const k of ["@id", "osm_id", "id"]) {
        if (props[k] != null) {
            const n = Number(String(props[k]).trim());
            if (!Number.isNaN(n)) return n;
        }
    }
    throw new Error("OSM id puuttuu (@id/osm_id/id). Aja osmium export -a id,type.");
}

function pickLevel(props) {
    const al = String(props.admin_level ?? "").trim();
    // pidetään vain 2,4,8
    return ["2", "3", "4", "5", "6", "7", "8"].includes(al) ? Number(al) : null;
}

function extractNamesAll(props) {
    const names = {};
    if (typeof props.name === "string" && props.name.trim()) names.name = props.name.trim();
    for (const [k, v] of Object.entries(props)) {
        if (typeof k === "string" && typeof v === "string" && k.startsWith("name:") && v.trim()) {
            const lang = k.split(":", 2)[1];
            names[lang] = v.trim();
        }
    }
    return names;
}

function chooseDisplayName(ns) {
    return ns.en || ns.fi || ns.name || (Object.values(ns)[0] ?? null);
}

// Pieni probe sen tarkistamiseksi onko tiedosto GeoJSON (sisältää "features")
async function looksLikeGeoJSON(filePath) {
    return new Promise((resolve) => {
        const rs = fs.createReadStream(filePath, { encoding: "utf8", highWaterMark: 4096 });
        let buf = "";
        rs.on("data", (chunk) => {
            buf += chunk;
            if (buf.length >= 4096) {
                rs.destroy();
                resolve(buf.includes('"features"'));
            }
        });
        rs.on("close", () => resolve(buf.includes('"features"')));
        rs.on("error", () => resolve(filePath.toLowerCase().endsWith(".geojson")));
    });
}

// ---------------- Admin-map (pieni JSON) ----------------
function parseAdminsMap(filePath) {
    const data = JSON.parse(fs.readFileSync(filePath, "utf8"));
    if (typeof data !== "object" || data === null || Array.isArray(data)) {
        throw new Error("admins-json ei ole dict.");
    }
    const out = {};
    for (const [k, vRaw] of Object.entries(data)) {
        if (typeof vRaw !== "object" || vRaw === null) continue;
        let v = { ...vRaw };
        if (v.level == null || v.id == null) {
            try {
                const [lvlStr, oidStr] = k.split("_");
                const lvl = Number(lvlStr.replace(/^l/, ""));
                const oid = Number(oidStr);
                if (!Number.isNaN(lvl) && !Number.isNaN(oid)) v = { level: lvl, id: oid, ...v };
            } catch { continue; }
        }
        out[k] = v;
    }
    return out;
}

// ---------------- GeoJSON streaming → Firestore ----------------
async function importGeoJSONStream(filePath, db, collectionName, dryRun = false) {
    return new Promise((resolve, reject) => {
        const col = db.collection(collectionName);

        let batch = db.batch();
        let inBatch = 0;
        let total = 0;
        let matched = 0;

        const pipeline = chain([
            fs.createReadStream(filePath),
            parser(),
            pick({ filter: "features" }),
            streamArray()
        ]);

        function commitNow(pauseable) {
            if (dryRun || inBatch === 0) return Promise.resolve();
            if (pauseable) pauseable.pause();
            const committing = inBatch;
            return batch.commit()
                .then(() => {
                    console.log(`Committed ${committing} docs (total ${total})`);
                    batch = db.batch();
                    inBatch = 0;
                })
                .finally(() => { if (pauseable) pauseable.resume(); });
        }

        pipeline.on("data", ({ value: feat }) => {
            total++;
            const props = feat?.properties || {};
            const geom = feat?.geometry || {};

            if (props["@type"] !== "relation") return;
            if (geom.type !== "Polygon" && geom.type !== "MultiPolygon") return;

            const level = pickLevel(props);
            if (level == null) return;

            let oid;
            try { oid = pickId(props); } catch { return; }

            const names = extractNamesAll(props);
            const disp = chooseDisplayName(names);
            const docId = `l${level}_${oid}`;

            matched++;

            if (!dryRun) {
                const ref = col.doc(docId);

                // Rakennetaan merge-payload pistemuotoisilla kentillä, jotta nimet täydentyvät
                const updateObj = { level, id: oid };
                if (disp) updateObj.displayName = disp;
                for (const [k, v] of Object.entries(names)) {
                    // Esim. names.en = "Finland", names.name = "Suomi"
                    updateObj[`names.${k}`] = v;
                }

                batch.set(ref, updateObj, { merge: true });
                inBatch++;

                if (inBatch >= BATCH_LIMIT) {
                    pipeline.pause();
                    commitNow(pipeline).catch(reject);
                }
            }

            if (total % 10000 === 0) {
                console.log(`Scanned ${total} features, eligible ${matched}, inBatch ${inBatch}`);
            }
        });

        pipeline.on("end", async () => {
            try {
                await commitNow(null);
                console.log(`Stream finished. Scanned ${total} features, eligible ${matched}.`);
                resolve(matched);
            } catch (e) {
                reject(e);
            }
        });

        pipeline.on("error", (e) => reject(e));
    });
}

// ---------------- Pieni JSON → Firestore ----------------
async function writeInBatches(db, collectionName, items, dryRun = false) {
    const col = db.collection(collectionName);
    let count = 0;
    let inBatch = 0;
    let batch = db.batch();

    for (const [docId, payload] of Object.entries(items)) {
        if (!dryRun) {
            const ref = col.doc(docId);

            // Jos admin-mapissa on valmiiksi names-objekti, kirjoitetaan se pisteillä mergeä varten
            const upd = {};
            for (const [k, v] of Object.entries(payload)) {
                if (k === "names" && v && typeof v === "object") {
                    for (const [nk, nv] of Object.entries(v)) {
                        upd[`names.${nk}`] = nv;
                    }
                } else {
                    upd[k] = v;
                }
            }

            batch.set(ref, upd, { merge: true });
            inBatch++;
        }
        count++;

        if (inBatch >= BATCH_LIMIT) {
            await batch.commit();
            console.log(`Committed ${inBatch} docs (total ${count})`);
            batch = db.batch();
            inBatch = 0;
        }
    }
    if (!dryRun && inBatch > 0) {
        await batch.commit();
        console.log(`Committed ${inBatch} docs (total ${count})`);
    }
    return count;
}

// ---------------- Main ----------------
(async function main() {
    try {
        const db = initFirestore(args.project, args.creds);

        const isGeo = await looksLikeGeoJSON(args.input);
        let total;

        if (isGeo) {
            console.log("Detected GeoJSON (streaming import) …");
            total = await importGeoJSONStream(args.input, db, "admin_areas", !!args.dryRun);
        } else {
            console.log("Detected small JSON map (buffered import) …");
            const docs = parseAdminsMap(args.input);
            console.log(`Docs to import: ${Object.keys(docs).length}`);
            total = await writeInBatches(db, "admin_areas", docs, !!args.dryRun);
        }

        console.log(`${args.dryRun ? "DRY RUN - " : ""}Imported/processed ${total} docs into 'admin_areas/'`);
        process.exit(0);
    } catch (err) {
        console.error("ERROR:", err && err.message ? err.message : err);
        process.exit(2);
    }
})();
